File: Tasks/runners/StrategyTaskRunner.cs
using Lab10.containers;
using Lab10.factory;
using Task = Lab10.domain.Task;

namespace Lab10.runners;

public class StrategyTaskRunner : AbstractTaskRunner
{
    //Clasa StrategyTaskRunner extinde clasa AbstractTaskRunner
    //Folosit pentru a alege strategia pentru gestionarea containerelor
    
    private Container container;
    
    //Constructor
    public StrategyTaskRunner(ContainerStrategy strategy)
    {
        container = TaskContainerFactory.getInstance().createContainer(strategy);
        if (container == null)
        {
            throw new InvalidOperationException("Failed to create a container.");
        }
    }
    
    //Suprascrie metoda executeOneTask din interfata TaskRunner
    public override void executeOneTask()
    {
        if (hasTask())
        {
            Task task = container.remove();
            if (task != null)
            {
                task.execute();
            }
        }
    }
    
    //Suprascrie metoda hasTask din interfata TaskRunner
    public override bool hasTask()
    {
        return !container.isEmpty();
    }
    
    //Suprascrie metoda executeAll din interfata TaskRunner
    public override void executeAll()
    {
        while (hasTask())
        {
            executeOneTask();
        }
    }
    
    //Suprascrie metoda addTask din interfata TaskRunner
    public override void addTask(Task task)
    {
        container.add(task);
    }
}

File: Tasks/src/decorator/StrategyTaskRunner.java
package decorator;

import factory.ContainerFactory;
import model.Container;
import model.Task;
import factory.TaskContainerFactory;
import factory.Strategy;

public class StrategyTaskRunner extends AbstractTaskRunner {
    //Implementeaza interfata TaskRunner

    private Container container;

    //Constructor
    public StrategyTaskRunner(Strategy strategy) {
        container=TaskContainerFactory.getInstance().createContainer(strategy);
    }

    //Suprascrie metoda executeOneTask() din interfata TaskRunner
    @Override
    public void executeOneTask() {
        if(!container.isEmpty()) {
            Task task = container.remove();
            if(task != null){
                task.execute();
            }
        }
    }

    //Suprascrie metoda executeAll() din interfata TaskRunner
    @Override
    public void executeAll() {
        while(hasTask()){
            executeOneTask();
        }
    }

    //Suprascrie metoda addTask() din interfata TaskRunner
    @Override
    public void addTask(Task t) {
        container.add(t);
    }

    //Suprascrie metoda hasTask() din interfata TaskRunner
    @Override
    public boolean hasTask() {
        return !container.isEmpty();
    }
}
